packages/mdc-menu-surface/foundation.ts

```diff
@@ -0,0 +1,500 @@
+  private getAutoLayoutMeasurements_(): AutoLayoutMeasurements {
+    let anchorRect = this.adapter_.getAnchorDimensions();
+    const bodySize = this.adapter_.getBodyDimensions();
+    const viewportSize = this.adapter_.getWindowDimensions();
+    const windowScroll = this.adapter_.getWindowScroll();
+
+    if (!anchorRect) {
+      anchorRect = {
+        top: this.position_.y,
+        right: this.position_.x,
+        bottom: this.position_.y,
+        left: this.position_.x,
+        width: 0,
+        height: 0,
+      };
+    }
+
+    return {
+      bodySize,
+      viewportSize,
+      windowScroll,
+      viewportDistance: {
+        top: anchorRect.top,
+        right: viewportSize.width - anchorRect.right,
+        left: anchorRect.left,
+        bottom: viewportSize.height - anchorRect.bottom,
+      },
+      anchorSize: anchorRect,
+      surfaceSize: this.dimensions_,
+    };
+  }
+
+  /**
+   * Computes the corner of the anchor from which to animate and position the menu surface.
+   */
+  private getOriginCorner_(): Corner {
+    // Defaults: open from the top left.
+    let corner = Corner.TOP_LEFT;
+
+    const {viewportDistance, anchorSize, surfaceSize} = this.measurements_;
+
+    const isBottomAligned = Boolean(this.anchorCorner_ & CornerBit.BOTTOM);
+    const availableTop = isBottomAligned ? viewportDistance.top + anchorSize.height + this.anchorMargin_.bottom
+      : viewportDistance.top + this.anchorMargin_.top;
+    const availableBottom = isBottomAligned ? viewportDistance.bottom - this.anchorMargin_.bottom
+      : viewportDistance.bottom + anchorSize.height - this.anchorMargin_.top;
+
+    const topOverflow = surfaceSize.height - availableTop;
+    const bottomOverflow = surfaceSize.height - availableBottom;
+    if (bottomOverflow > 0 && topOverflow < bottomOverflow) {
+      corner |= CornerBit.BOTTOM;
+    }
+
+    const isRtl = this.adapter_.isRtl();
+    const isFlipRtl = Boolean(this.anchorCorner_ & CornerBit.FLIP_RTL);
+    const avoidHorizontalOverlap = Boolean(this.anchorCorner_ & CornerBit.RIGHT);
+    const isAlignedRight = (avoidHorizontalOverlap && !isRtl) ||
+      (!avoidHorizontalOverlap && isFlipRtl && isRtl);
+    const availableLeft = isAlignedRight ? viewportDistance.left + anchorSize.width + this.anchorMargin_.right :
+      viewportDistance.left + this.anchorMargin_.left;
+    const availableRight = isAlignedRight ? viewportDistance.right - this.anchorMargin_.right :
+      viewportDistance.right + anchorSize.width - this.anchorMargin_.left;
+
+    const leftOverflow = surfaceSize.width - availableLeft;
+    const rightOverflow = surfaceSize.width - availableRight;
+
+    if ((leftOverflow < 0 && isAlignedRight && isRtl) ||
+        (avoidHorizontalOverlap && !isAlignedRight && leftOverflow < 0) ||
+        (rightOverflow > 0 && leftOverflow < rightOverflow)) {
+      corner |= CornerBit.RIGHT;
+    }
+
+    return corner;
+  }
+
+  /**
+   * @param corner Origin corner of the menu surface.
```

Comment:
remove closure comments

Reply:
This is actually [TypeDoc](https://typedoc.org/guides/doccomments/) syntax 😀